Allow lookup of secrets by name.

joaogbcravo opened the pull request:
My second attempt to merge this feature.. 1 year after :) 

Allows getting Bitwarden secrets using the name of the Bitwarden Secret and its location (in a collection) instead of an ID.

Also, reduce duplication of code and improve some logic.

joaogbcravo commented on src/bitwardenCrdOperator.py:
```diff
@@ -19,6 +19,7 @@ def bitwarden_signin(logger, **kwargs):
         logger.info("BW_HOST not set. Assuming SaaS installation")
     command_wrapper(logger, "login --apikey")
     unlock_bw(logger)
+    sync_bw(logger)
```
For some reason, I got some race condition situation where the vault was not synced. I guess it's important to do this here.

joaogbcravo commented:
@Lerentis any thoughts on this one?

Lerentis commented:
@joaogbcravo i will have a look over the weekend 👍

Lerentis commented on src/template.py:
```diff
@@ -12,11 +12,38 @@ def render_template(logger, template):
+
+    owner_references = [{
+        "apiVersion": f"{body.get('apiVersion')}",
+        "blockOwnerDeletion": True,
```
what is the intention behind `blockOwnerDeletion` and `controller`?

joaogbcravo replied:
I don't want the [owner to be deleted unless this is also deleted](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/object-meta/#System:~:text=time%20package%20offers.-,ownerReferences,-(%5B%5DOwnerReference)).

And I also want the generated secret to be deleted if the bitwarden-templates.lerentis object is deleted.

Lerentis commented on src/utils/utils.py:
```diff
@@ -57,8 +65,9 @@ def unlock_bw(logger):
+    final_command = f"bw {response_flag} {command}"
     sp = subprocess.Popen(
-        [f"bw {response_flag} {command}"],
+        [final_command],
```
why are you wrapping this again in a variable?

Lerentis commented on src/utils/utils.py:
```diff
@@ -31,12 +41,10 @@ def _sync(logger):
         logger.debug("Running scheduled sync")
-        status_output = _sync(logger)
-        logger.info(f"Sync successful {status_output}")
+        _sync(logger)
```
any reason you removed the log entry afterwards?

Lerentis commented on src/lookups/bitwarden_lookup.py:
```diff
@@ -1,4 +1,5 @@
-from utils.utils import get_secret_from_bitwarden, get_attachment, parse_fields_scope, parse_login_scope
+from utils.utils import get_secret_from_bitwarden, get_collection_secret_from_bitwarden, \
+                        get_attachment, parse_fields_scope, parse_login_scope
```
missing `get_secret_with_id_from_bitwarden`

Lerentis reviewed (changes requested):
You missed the adaptation of https://github.com/Lerentis/bitwarden-crd-operator/blob/main/src/dockerlogin.py#L128 to the new method header of `get_secret_from_bitwarden`.
other than that and the missing import there are just a few questions to answer.
please also add the following: 
* increase the crd beta version while still serving the old version (you can drop v1beta4 if you like)
* write a changelog 
* bump the chart and the image version to a new minor

joaogbcravo commented:
> write a changelog

On the chart.yaml or somewhere else?

Lerentis commented:
> > write a changelog
> 
> On the chart.yaml or somewhere else?

yep it will be listed on artifacthub: https://artifacthub.io/packages/helm/lerentis/bitwarden-crd-operator?modal=changelog

joaogbcravo commented:
> > > write a changelog
> > 
> > 
> > On the chart.yaml or somewhere else?
> 
> yep it will be listed on artifacthub: https://artifacthub.io/packages/helm/lerentis/bitwarden-crd-operator?modal=changelog

Addressed your comments on the last commit.

Lerentis commented:
@joaogbcravo does this code work for you? 
in my tests the secrets are no longer created by the operator. looks like there is a logical error in the chain of events.
```
  Normal   Logging  61s   kopf  Secret default/test-secret has been created
  Normal   Logging  61s   kopf  Handler 'create_managed_secret' succeeded.
  Warning  Logging  61s   kopf  Could not update secret default/test-secret!
```

```bash
kubectl get secrets -n default
No resources found in default namespace.
```

my test setup:

kind v0.22.0 go1.22.0 linux/amd64
Server Version: v1.29.2
```bash
kind create cluster
checkout your branch
docker build -t test .
kind load docker-image test
helm upgrade --install -n bitwarden-crd-operator bitwarden-crd-operator -f charts/bitwarden-crd-operator/myvalues.yaml charts/bitwarden-crd-operator/ --set image.tag=latest --set image.repository=test --create-namespace
kubectl apply -f example*.yaml
```


other than that the example is still missing a required name

joaogbcravo commented:
@Lerentis Friendly ping :)
Anything I can do to help you?

Lerentis commented:
Hi @joaogbcravo ,
i tried a couple of configurations over the weekend (kind, minikube and a kops created cluster on hetzner) but i can not get this code to work. it just simply silently fails to create the managed secret, while the logs just state that it does.
frustratingly i can not even pin point the reason why the secret is not created. in an upgrade situation (current main to your PR) all managed secrets were deleted from the cluster.
The code on main works without issues, so this is somewhere in the execution flow of this change.

in this state i can not merge this PR. i dont know if k3s is doing something different but if it does it is not api compatible with kopf i fear.
If you can provide working code with kind/minikube for testing i reconsider this but for now i will not continue to debug this

joaogbcravo commented:
Just tried with Minikube:

```
$ minikube version
minikube version: v1.29.0
commit: ddac20b4b34a9c8c857fc602203b6ba2679794d3
```

And I managed to replicate it!
The problem is I managed to replicate it once, and not anymore.
I wonder if this is some kind of race condition. 

For you, does it happen all the time?

Can you maybe do a last debug for me? Can you change the code to raise the exception on [this block](https://github.com/Lerentis/bitwarden-crd-operator/blob/33ac29f959c351a1d62460d12fff7222e5b54471/src/kv.py#L132) so that we can understand the root of the problem?

joaogbcravo commented:
I think I can replicate it if I delete/create the secret. I managed to get this error... I will look on it

```
  Normal  Logging  39s   kopf  Already unlocked
  Normal  Logging  39s   kopf  Already unlocked
  Normal  Logging  38s   kopf  Looking up secret with ID: db4a5afe-21f0-4bb1-844c-b11a01150cb0
  Normal  Logging  37s   kopf  Looking up secret with ID: db4a5afe-21f0-4bb1-844c-b11a01150cb0
  Normal  Logging  38s   kopf  Sync successful {'success': True, 'data': {'noColor': False, 'object': 'message', 'title': 'Syncing complete.', 'message': None}}
  Normal  Logging  37s   kopf  Sync successful {'success': True, 'data': {'noColor': False, 'object': 'message', 'title': 'Syncing complete.', 'message': None}}
  Normal  Logging  36s   kopf  Handler 'create_managed_secret' succeeded.
  Error   Logging  37s   kopf  Timer 'update_managed_secret' failed with an exception. Will retry.
Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/kopf/_core/actions/execution.py", line 276, in execute_handler_once
    result = await invoke_handler(
             ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/kopf/_core/actions/execution.py", line 371, in invoke_handler
    result = await invocation.invoke(
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/kopf/_... 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'X-Kubernetes-Pf-Flowschema-Uid': '6a8618f6-5796-4a95-8198-2ddea5c5f227', 'X-Kubernetes-Pf-Prioritylevel-Uid': 'fa71be9c-91f9-434e-a7cc-515a638e33f0', 'Date': 'Tue, 12 Mar 2024 19:34:03 GMT', 'Content-Length': '196'})
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"secrets \"test-secret\" not found","reason":"NotFound","details":{"name":"test-secret","kind":"secrets"},"code":404}
  Normal   Logging  36s  kopf  Secret default/test-secret has been created
  Warning  Logging  37s  kopf  Could not update secret default/test-secret!
  Normal   Logging  36s  kopf  Creation is processed: 1 succeeded; 0 failed.

```

joaogbcravo commented:
@Lerentis I managed to replicate the problem (same exception) also with the main branch code. 

Can you confirm the exception you are getting is the same or other?

I used skaffold :)

Lerentis commented:
@joaogbcravo for me this is reproducible every time and does not happen on main. if you want to we can have a debug session together via discord? 
send me your details to `lerentis at uploadfilter24 dot eu`

Lerentis commented:
@joaogbcravo ping. Did not get a mail from you

joaogbcravo commented:
Hi @Lerentis I sent it yesterday!

Lerentis commented:
closing due to inactivity
